src/codegen/toplevel_definitions.jl

```diff
@@ -191,15 +245,41 @@ function translate(io, rp::ResolvedProtoFile, file_map::Dict{String,ResolvedProt
             len += length(name)
         end
     end
-    !isempty(p.cyclic_definitions) && println(io, "\n\n# Abstract types to help resolve mutually recursive definitions")
+
+    println(io)
+
+    if options.parametrize_oneofs
+        for name in @view(p.sorted_definitions[1:end-ncyclic])
+            def = p.definitions[name]
+            if def isa MessageType && def.has_oneof_field && def.is_self_referential[]
+                println(io, "abstract type ", abstract_type_name(name), options.common_abstract_type ? " <: AbstractProtoBufMessage" : "", " end")
+            end
```

Comment:
this is very similar logic to in `generate_struct` (up around L50 ish), i wonder if they can/should be refactored out (but i don't immediately see a nice way)

Reply:
I've added a helper function to use at the two locations 👍